test/cql-pytest/test_unset.py

```diff
@@ -0,0 +1,167 @@
+
+#############################################################################
+# Tests for the UNSET_VALUE value introduced in CQL version 4. Unset values
+# can be bound to variables, which cause certain CQL assignments to be skipped,
+# and may have other effects on other requests.
+
+import pytest
+from util import new_test_table, unique_key_int
+from cassandra.query import UNSET_VALUE
+from cassandra.cluster import NoHostAvailable
+from cassandra.protocol import InvalidRequest
+
+@pytest.fixture(scope="module")
+def table1(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "p int PRIMARY KEY, a int, b int, c int, li list<int>") as table:
+        yield table
+
+@pytest.fixture(scope="module")
+def table2(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "p int, c int, PRIMARY KEY (p, c)") as table:
+        yield table
+
+# A basic test that in a prepared statement with three assignments, one
+# bound by an UNSET_VALUE is simply not done, but the other ones are.
+# Try all 2^3 combinations of a 3 column updates with each one set to either
+# a real value or an UNSET_VALUE.
+def test_update_unset_value_basic(cql, table1):
+    p = unique_key_int()
+    stmt = cql.prepare(f'UPDATE {table1} SET a=?, b=?, c=? WHERE p={p}')
+    a = 1
+    b = 2
+    c = 3
+    cql.execute(stmt, [a, b, c])
+    assert [(a, b, c)] == list(cql.execute(f'SELECT a,b,c FROM {table1} WHERE p = {p}'))
+    i = 4
+    for unset_a in [False, True]:
+        for unset_b in [False, True]:
+            for unset_c in [False, True]:
+                if unset_a:
+                    newa = UNSET_VALUE
+                else:
+                    newa = i
+                    a = i
+                    i += 1
+                if unset_b:
+                    newb = UNSET_VALUE
+                else:
+                    newb = i
+                    b = i
+                    i += 1
+                if unset_c:
+                    newc = UNSET_VALUE
+                else:
+                    newc = i
+                    c = i
+                    i += 1
+                cql.execute(stmt, [newa, newb, newc])
+                assert [(a, b, c)] == list(cql.execute(f'SELECT a,b,c FROM {table1} WHERE p = {p}'))
+
+# The expression "SET a=?" is skipped if the bound value is UNSET_VALUE.
+# But what if it is a more complex expression, like "SET li=li+?" (list append)
+# or "SET a=(int)?+1" (arithmetic expression on the bind variable)? Does
+# the SET also get skipped? Unfortunately, the answer is no: Cassandra
+# refuses or fails to evaluate expressions involving an UNSET_VALUE, and
+# it causes the whole write request to fail instead of parts of it being
+# skipped. Noticing this, we decided (see discussion in pull request #12517)
+# that this will become a rule: expressions involving UNSET_VALUE will
+# always be refused. The following tests verify this in a few cases.
+
+@pytest.mark.xfail(reason="append with unset doesn't produce error")
```

Comment:
This isn't xfail, it's expected to pass. Append is a primitive operation even if it is specified as a complex expression.

Reply:
This is a good point. As I noted below, the simlar **counter**  case (also a primitive operation) is explicitly described in the protocol documentation as something that needs to be skipped - not error out - so the list append should probably do the same thing. This means I'll have to change this **xfail** to **cassandra_bug** - i.e., something that after deliberation we decided that Cassandra is doing the wrong thing and we will do something else - deliberately.

I'll send a v2 doing this change and improving the comments to explain what we discussed here.

Reply:
Ah, I thought Cassandra was in fact honoring the unset here.

So it's somewhat accidental, I didn't intend to go beyond Cassandra here. But note my patch did not change the behavior, AFAIK.